crates/walrus-service/src/node/blob_sync.rs

```diff
@@ -743,6 +648,147 @@ impl BlobSynchronizer {
+            )
+        })
+        .into_iter()
+        .map(|shard| {
+            blob_synchronizer
+                .clone()
+                .recover_slivers_for_shard(shared_metadata.clone(), shard)
+        });
+
+    let mut futures_with_permits = stream::iter(futures_iter).then(move |future| {
+        let permits = sliver_permits.clone();
+
+        // We use a future to get the permit. Only then is the future returned from the stream
+        // to be awaited.
+        #[allow(clippy::async_yields_async)]
+        async move {
+            let claimed_permit = permits
+                .acquire_owned()
+                .await
+                .expect("semaphore should not been dropped");
+            // Attach the permit to the future, so that it is held until the future completes.
+            future.map(|result| (result, claimed_permit))
+        }
+    });
+    let mut futures_with_permits = std::pin::pin!(futures_with_permits);
+    // When stored in a JoinSet, tasks are aborted on drop.
+    let mut pending_tasks = JoinSet::new();
+
+    loop {
+        tokio::select! {
+            biased;
+
+            Some(future) = futures_with_permits.next() => {
+                // Add the permit and the future to the list of pending futures
+                pending_tasks.spawn(future);
+            }
+            Some(join_result) = pending_tasks.join_next() => {
+                match join_result {
+                    Ok((Err(RecoverSliverError::Inconsistent(inconsistency_proof)), _permit)) =>
+                    {
+                        tracing::warn!("received an inconsistency proof");
+                        // No need to recover other slivers, sync the proof and return
+                        blob_synchronizer.sync_inconsistency_proof(&inconsistency_proof)
+                            .observe_future(histograms.clone(),
+                                            labels_from_inconsistency_sync_result)
+                            .await;
+                        break;
+                    }
+                    Ok((Err(RecoverSliverError::Database(err)), _)) => {
+                        match err {
+                            TypedStoreError::UnregisteredColumn(ref column) => {
+                                // It is possible that when the blob recovery is running, a shard
+                                // is removed from the storage, and therefore encounter
+                                // UnregisteredColumn error. In this case, we should not panic
+                                // and instead return an error to retry the blob sync.
+                                tracing::error!(
+                                    "database error during sliver sync: unregistered column {:?}",
+                                    column
+                                );
+                                return Err(RecoverSliverError::Database(err));
+                            }
```

Comment:
Here we need to special case CF not found error so that this error can be retried.